runtime/io/io.cpp

```diff
@@ -0,0 +1,389 @@
+    case ETIMEDOUT:
+      retBlock->h = getBlockHeaderForSymbol((((uint64_t)getTagForSymbolName(ERRTAG(ETIMEDOUT))) << 32) | 1);
+    case ECONNREFUSED:
+      retBlock->h = getBlockHeaderForSymbol((((uint64_t)getTagForSymbolName(ERRTAG(ECONNREFUSED))) << 32) | 1);
+    case EHOSTDOWN:
+      retBlock->h = getBlockHeaderForSymbol((((uint64_t)getTagForSymbolName(ERRTAG(EHOSTDOWN))) << 32) | 1);
+    case EHOSTUNREACH:
+      retBlock->h = getBlockHeaderForSymbol((((uint64_t)getTagForSymbolName(ERRTAG(EHOSTUNREACH))) << 32) | 1);
+    case ELOOP:
+      retBlock->h = getBlockHeaderForSymbol((((uint64_t)getTagForSymbolName(ERRTAG(ELOOP))) << 32) | 1);
+    case EOVERFLOW:
+      retBlock->h = getBlockHeaderForSymbol((((uint64_t)getTagForSymbolName(ERRTAG(EOVERFLOW))) << 32) | 1);
+    default:
+      retBlock = static_cast<block *>(koreResizeLastAlloc(retBlock, sizeof(block) + sizeof(uint64_t), sizeof(block)));
+      retBlock->h = getBlockHeaderForSymbol((uint64_t)getTagForSymbolName("Lbl'Hash'unkownIOError"));
+      mpz_t err;
+      mpz_init_set_si(err, errno);
+      mpz_ptr p = move_int(err);
+      memcpy(&(retBlock->children[0]), &p, sizeof(uint64_t));
+    }
+
+    return retBlock;
+  }
+
+  blockheader header_int() {
+    static blockheader header = {(uint64_t)-1};
+
+    if (header.hdr == -1) {
+      getBlockHeaderForSymbol((uint64_t)getTagForSymbolName("inj{SortInt, SortIOInt}"));
+    }
+
+    return header;
+  }
+
+  blockheader header_err() {
+    static blockheader header = {(uint64_t)-1};
+
+    if (header.hdr == -1) {
+      getBlockHeaderForSymbol((uint64_t)getTagForSymbolName("inj{SortIOError, SortKItem}"));
+    }
+
+    return header;
+  }
+
+  blockheader header_string() {
+    static blockheader header = {(uint64_t)-1};
+
+    if (header.hdr == -1) {
+      getBlockHeaderForSymbol((uint64_t)getTagForSymbolName("inj{SortString, SortIOString}"));
+    }
+
+    return header;
+  }
+
+  char * getTerminatedString(string * str) {
+    int length = len(str);
+    char * buf = (char *) malloc(sizeof(KCHAR) * (length + 1));
+    strncpy(buf, str->data, length);
+    buf[length] = '\0';
+    return buf;
+  }
+
+  block * hook_IO_open(string * filename, string * control) {
+    int flags = 0;
+    int length = len(control);
+
+    switch (control->data[0]) {
+      case 'r':
+        flags = O_RDONLY;
+      case 'w':
+        flags = O_WRONLY | O_TRUNC | O_CREAT;
+      case 'a':
+        flags = O_WRONLY | O_APPEND | O_CREAT;
+      default:
+        flags = O_RDONLY;
+    }
+
+    if ((length > 1 && control->data[1] == '+') || (length > 2 && control->data[2] == '+')) {
+      flags &= (~O_WRONLY) & ~(O_RDONLY);
+      flags |= O_RDWR;
+    }
+
+    mpz_t result;
+    char * f = getTerminatedString(filename);
+    int fd = open(f, flags, S_IRUSR | S_IWUSR | S_IRGRP | S_IWGRP | S_IROTH);
+
+    block * retBlock = static_cast<block *>(koreAlloc(sizeof(block) + sizeof(uint64_t)));
```

Comment:
I don't think it's going to be acceptable to have allocation function calls in between the call to open and reading the value of errno if open failed. koreAlloc can in some cases call malloc which presumably can call all kinds of system calls that reset errno.